move nsis

RoDuth opened the pull request.

mfrasca commented on nsis/build-multiuser.nsi:
```diff
@@ -49,9 +49,9 @@
-!addincludedir "..\data\nsis\Include\"
+!addplugindir /x86-ansi "Plugins\x86-ansi\"
+!addplugindir /x86-unicode "Plugins\x86-unicode\"
+!addincludedir "Include\"
```
are you sure this is correct?  shouldn't it be `..\nsis\<etc>`?

RoDuth replied:
I changed the location of the script to `nsis` also so relative is no longer one step back.  It is all together now.  This also means the installer ends up in the `nsis` directory when all is done.  I've changed this in the docs also.  It runs fine for me and produces a functional installer as far as I can tell in my win 7 VM.  If anything just check it reads right in the docs.

mfrasca replied:
thanks, much better (that file alone is 37% of the current repository size).

I think I still would rather see the `build-win` script in the `scripts` directory, and that after running it the generated executable ends in the `dist` directory: not for nothing the `dist` directory is ignored, it's because it only contains generated files.  store it there, and you won't make the mistake of committing it again.  store a script among all other scripts, and chances are that someone else will find it even without reading the relevant documentation section.

but after all, I can do this myself, when I come to it.

RoDuth commented:
sorry thought I had...

mfrasca commented on nsis/build_win.bat:
```diff
@@ -71,7 +72,7 @@ python setup.py nsis
 
 :SKIP_NSIS
-copy scripts\win_gtk.bat dist
+copy nsis\win_gtk.bat dist
```
why copying a script to the `dist` directory?

mfrasca commented on nsis/build_win.bat:
```diff
@@ -6,19 +6,20 @@ REM only other argument proccessed must be a pathname to a virtualenv)
 ) else (
-  set venv="%HOMEDRIVE%%HOMEPATH%\.virtualenvs\ghi2exe"
+  for /f %%i in ('git rev-parse --abbrev-ref HEAD') do (set branch=%%i)
+  set venv="%HOMEDRIVE%%HOMEPATH%\.virtualenvs\%branch%-2exe"
```
:+1:

mfrasca commented:
wow, this is going to be difficult … the file you committed, it's now in the history.  it wasn't enough removing it, it's in the history… now I'm not sure how to remove it, it is possible, it is NOT possible to do it clean, I have to do it quickly, before too many people clone the repository or update it.

mfrasca commented:
@RoDuth , would you mind doing this again? I could not find any other way to remove your executable from history, but forcing push of a repository where history did not contain this merge.
**you cannot just "remove" the file, you have undo the action of adding it, it's not simple.**
https://stackoverflow.com/questions/307828/completely-remove-file-from-all-git-repository-commit-history#308684
or you redo everything from scratch, that's probably the easiest.
I'm afraid that you added the file, then removed it, then added it again and then removed on my request.  anyway, once written in the history, it's occupying space in the repository.
as I said, putting generated files in `dist` is the way to avoid such mistakes.

RoDuth commented:
@mfrasca sorry, I'm moving house at the moment and having trouble getting an internet connection put on at the new place. Have been without internet for 2 days and will only have occasional connection until we sort it out.

So to your comments above:

> I think I still would rather see the build-win script in the scripts directory, and that after running it the generated executable ends in the dist directory: not for nothing the dist directory is ignored, it's because it only contains generated files. store it there, and you won't make the mistake of committing it again. store a script among all other scripts, and chances are that someone else will find it even without reading the relevant documentation section.

I actually prefer `build-win.bat` here, in `nsis`, with everything in the one place.  `dist` , being the output of py2exe is a full self contained version of the software and I think it not best to put the installer in there also.  As it is now the installer ends up in `nsis` which makes sense to me.  After realising that I had committed it I just added `*setup.exe` to `.gitignore` no more problems no mater where I put it.  Placing it in the `dist` folder could cause issues for people who then also want to use the output in `dist` on a usb or the like (placing `win_gtk.bat` manually - see below).  Also by leaving it in `nsis` (and gitignoring it) allows you to keep copies of any previous version of the installer you have built as `dist` is deleted by `python setup.py clean`, which is called each time `build-win.bat` is called.

> why copying a script to the dist directory?

from step 7 in the documentation:

> The `dist` folder will now contain a full working copy of the software in a frozen, self contained state, that can be transferred however you like and will work in place. (e.g. placed on a USB flash drive for demonstration purposes or copied manually to `C:\Program Files` with a shortcut created on the desktop). To start ghini.desktop double click `ghini.exe` in explorer (or create a shortcut to it). If you have issues with the UI not displaying correctly you need to run the script `win_gtk.bat` from the `dist` folder to set up paths to the GTK components correctly. (Running `build_win /e` will place this script in the dist folder for you or you can copy it from the `nsis` folder yourself.) You will only need to run this once each time the location of the folder changes. Thereafter ghini.exe will run as expected.

Basically, this only happens when calling `build_win /e`, the `win_gtk.bat` script is needed to set up configuration files for gtk to work correctly.  You'll need to run it each time the folder is moved (the nsis installer takes care of this when installing) so if you want a demonstration version on a usb just run that batch file first each new computer you plug into (or you could just copy the folder somewhere on their computer, run the script and make a shortcut on their desktop).  From then on `ghini.exe` will work.  There are limitations to this method compared to the nsis installer but for a quick way to show people at a conference or the like it is very handy.

>  would you mind doing this again?

I'm assuming you are asking me to `git reset <sha-before-the-changes>` > ` git add .` (with my gitignore fixed to stop *setup.exe) > `git commit` > `git push --force` or something along those lines?
Something I hate doing but I guess I could in this situation.  Or to use the rebase formula from SO, which I'm not sure about.

mfrasca commented:
I restored the repository to the situation before merging your branch, so I'm afraid you need to restore history yourself: I very likely made a change that won't be recognized by your git repository.
as said, I suggest you redo things from scratch, not that you correct the mistake, much easier.

mfrasca commented:
> I'm moving house at the moment and having trouble getting an internet connection put on at the new place

re-reading all this, actually I don't think there's anything else to do here from your side at least, I will need reviewing, and making sure that I will again understand the organization in a couple of months or years, when I know for sure I will have forgotten everything.

RoDuth commented:
> I have not yet reviewed your documentation

Almost everything you have asked, for over 6 months now, and I have often had to repeat myself, sometimes multiple times, could have been answered by you reading the documentation.  Why not read it?  I really don't understand.

> I'm quite happy that we can [place the content of a single directory] on a USB flash drive for demonstration purposes, I'm not yet sure how you did it, and I need to have a look, and while I hope you clearly separated this from the rest of the dist directory, from your description it sounds like you did not.
I need to be in control of the definition of the roles of each directory. I am not particularly happy with your interpretations, which you seem to implement without discussion.

How did I do it? Py2exe does almost all of it.  @brettatoms did most of the work and I just updated it and made a few fairly minor improvements (a git blame will show this) so there was no interpretations to discuss, I changed nothing.  As @takluyver has mentioned 'The trickiest bit is getting the necessary files for pygtk in the right place.' so, as py2exe fails to get it right, we ask py2exe to ignore it and do it separately.  Brett's approach seems similar to everything I saw in the py2exe docs and other sources I looked at so I didn't change it beyond a few minor improvements.

As for 'separated this from the rest of the dist directory' I would think you should run `setup.py clean` so I don't see any issue.  But if you like you can add something like `"dist_dir": "exe",` to `py2exe_options` in `setup.py` and I would also include that directory in `setup.py clean`.

> `nsis`, I'm also not sure I can describe what it contains. isn't it a local copy of nsis files? why do we keep them in the repository? what about licenses?

From memory: Licenses for the nsis plugins are all fine (gpl compatible). Would have been better able to reply to this 6 months ago as I know I spent some time looking into it at the time.  Whole versions are included for this reason.

We keep them in the repository to avoid having to source the correct versions elsewhere.  They come from differing sources with varying amounts of activity.  The main one that has had braking changes in the past is NsisMultiUser.  To use them you need to extract them and place the relevant components in the correct place...  Or you just set them up once in the repo and point your nsis script to them and it works without having to go through any of that.

I feel that if you do wish to split out the nsis stuff then it should all be put in `nsis`.  After all everything I have now put there is only relevant to producing a windows executable and/or an installer.  I may have called it something more generic like `win` etc. but that isn't important.

> actually I don't think there's anything else I need you to do,

I'm having issues with internet connection and time at the moment (moving with a 2 year old isn't easy!) so I'm going to assume you will do what ever it is that you you want to do.

mfrasca commented:
>  Why not read it? I really don't understand.

Every time I try reading, I get more confused than I was before.  I need a step-by-step guide for system set-up, and a script that simply works after the system has been set up.  It might be that I just do not manage to read anything related to Windows, I don't know.  I remember reading and doing and missing steps and asking and hearing "oh, but you need that, too. (wasn't that clear).", it wasn't to me, and I'm sure I'm recalling incorrectly.  anyhow: don't worry, really, I guess it's just me who can't read Windows and native speakers.  thank you for all the effort.

RoDuth commented:
> a step-by-step guide for system set-up,

It really is, as stated in the docs:
1. Download and install git, Python 2.7 and PyGTK as outlined in the generic installation instructions.

2. Additionally, download and install NSIS v3.

> and a script that simply works

if you have followed these steps and have not done anything unusual, like install ghini globally, outside of a virtual environment for instance (your last issue).   It really will work.  I have repeated this same process many many times in windows 7, 8 and 10 VMs and on my own laptop install and others.  Can I also add that your `devinstall.bat` has issues that make it fail for unusual situations.  The most glaring one that catches me out every time is the fact that it fails if you have python 3 setup as your default python.  At least `build_win.bat` doesn't have that issue.  This is what makes the installer the way it is with py2exe etc. such a good way to distribute.  It doesn't rely on any dependencies and should not be affected by anything unusual about the users system.  This self contained setup is what windows users are used to. What if a user is cleaning up their system (as windows users seem to do all the time!) and removes git for instance.

I've rewritten the documentation twice now, I've included all the details you have asked for, I've tried it on not so computer savvy people and on people not so familiar with windows, give it another look.

mfrasca commented:
mentioning problems is the first step towards solving them.  as said in #109: apart from me not understanding that after running `build-win` you get a relocatable installed ghini.desktop in the `dist` directory, and from the fact that the distributable installer is put in `scripts`, all is quite fine.  I take in on myself to label things in a way that will let me find things back again in a couple of weeks, when I'll have forgotten every single detail.
thanks for all your work.
